From f3343389bd5d7e31114b7677d99bbd132232b100 Mon Sep 17 00:00:00 2001
From: Andrey Yepin <ayepin@google.com>
Date: Fri, 6 Mar 2026 14:47:10 -0800
Subject: [PATCH] Sanitize alternate intents.

Exclude ineligible alternate intents from cross-profile target
resolution.

Bug: 444443479
Test: manual testing with a test app
Test: test interactive session
Test: test Shareousel
Test: new STS tests
Flag: EXEMPT CVE_FIX
Change-Id: I749844b3d89105fe14411ed762de4d221472ff9f
---
 .../intentresolver/ChooserActivity.java       |  20 +-
 .../AlwaysTrueCrossProfileIntentsChecker.kt   |  36 ++++
 .../CrossProfileIntentsChecker.java           |   7 +-
 .../intentresolver/util/IntentUtils.kt        |  91 ++++++++-
 .../intentresolver/util/IntentUtilsTest.kt    | 173 ++++++++++++++++--
 5 files changed, 299 insertions(+), 28 deletions(-)
 create mode 100644 java/src/com/android/intentresolver/emptystate/AlwaysTrueCrossProfileIntentsChecker.kt

diff --git a/java/src/com/android/intentresolver/ChooserActivity.java b/java/src/com/android/intentresolver/ChooserActivity.java
index 85ee598f..9c700334 100644
--- a/java/src/com/android/intentresolver/ChooserActivity.java
+++ b/java/src/com/android/intentresolver/ChooserActivity.java
@@ -34,6 +34,7 @@ import static com.android.intentresolver.Flags.unselectFinalItem;
 import static com.android.intentresolver.ext.CreationExtrasExtKt.replaceDefaultArgs;
 import static com.android.intentresolver.profiles.MultiProfilePagerAdapter.PROFILE_PERSONAL;
 import static com.android.intentresolver.profiles.MultiProfilePagerAdapter.PROFILE_WORK;
+import static com.android.intentresolver.util.IntentUtils.prepareCrossProfileIntents;
 import static com.android.internal.util.LatencyTracker.ACTION_LOAD_SHARE_SHEET;
 
 import static java.util.Objects.requireNonNull;
@@ -110,6 +111,7 @@ import com.android.intentresolver.data.model.ChooserRequest;
 import com.android.intentresolver.data.repository.ActivityModelRepository;
 import com.android.intentresolver.data.repository.DevicePolicyResources;
 import com.android.intentresolver.domain.interactor.UserInteractor;
+import com.android.intentresolver.emptystate.AlwaysTrueCrossProfileIntentsChecker;
 import com.android.intentresolver.emptystate.CompositeEmptyStateProvider;
 import com.android.intentresolver.emptystate.CrossProfileIntentsChecker;
 import com.android.intentresolver.emptystate.EmptyStateProvider;
@@ -1201,7 +1203,7 @@ public class ChooserActivity extends Hilt_ChooserActivity implements
     // @NonFinalForTesting
     @VisibleForTesting
     protected CrossProfileIntentsChecker createCrossProfileIntentsChecker() {
-        return new CrossProfileIntentsChecker(getContentResolver());
+        return new AlwaysTrueCrossProfileIntentsChecker(getContentResolver());
     }
 
     protected final EmptyStateProvider createEmptyStateProvider(
@@ -1519,7 +1521,12 @@ public class ChooserActivity extends Hilt_ChooserActivity implements
             ChooserGridAdapter adapter = createChooserGridAdapter(
                     context,
                     isCrossProfile
-                            ? request.getCrossProfilePayloadIntents()
+                            ? prepareCrossProfileIntents(
+                                    getContentResolver(),
+                                    mRequest.getTargetIntent(),
+                                    request.getCrossProfilePayloadIntents(),
+                                    getEffectiveProfile(launchedAs).getPrimary().getHandle(),
+                                    getEffectiveProfile(profile).getPrimary().getHandle())
                             : request.getPayloadIntents(),
                     isCrossProfile ? null : initialIntentArray,
                     profile.getPrimary().getHandle()
@@ -1551,6 +1558,15 @@ public class ChooserActivity extends Hilt_ChooserActivity implements
                 maxTargetsPerRow);
     }
 
+    private Profile getEffectiveProfile(Profile profile) {
+        // When sharing into or out of Private profile, perform the check using the parent profile
+        // instead. (Hard-coded application of CROSS_PROFILE_CONTENT_SHARING_DELEGATE_FROM_PARENT)
+        // Assumption baked into design: "Personal" profile is the parent of all other profiles.
+        return profile.getType() == Profile.Type.PRIVATE
+                ? mProfiles.getPersonalProfile()
+                : profile;
+    }
+
     protected EmptyStateProvider createBlockerEmptyStateProvider() {
         return new NoCrossProfileEmptyStateProvider(
                 mProfiles,
diff --git a/java/src/com/android/intentresolver/emptystate/AlwaysTrueCrossProfileIntentsChecker.kt b/java/src/com/android/intentresolver/emptystate/AlwaysTrueCrossProfileIntentsChecker.kt
new file mode 100644
index 00000000..bd1203a3
--- /dev/null
+++ b/java/src/com/android/intentresolver/emptystate/AlwaysTrueCrossProfileIntentsChecker.kt
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2026 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.intentresolver.emptystate
+
+import android.annotation.UserIdInt
+import android.content.ContentResolver
+import android.content.Intent
+
+/**
+ * A [CrossProfileIntentsChecker] that always returns true, it is used to test the cross profile
+ * intent policy bypass. To be used in a case when the blocked cross-profile intents has been vetted
+ * elsewhere.
+ */
+class AlwaysTrueCrossProfileIntentsChecker(contentResolver: ContentResolver) :
+    CrossProfileIntentsChecker(contentResolver) {
+
+    override fun hasCrossProfileIntents(
+        intents: List<Intent>,
+        @UserIdInt source: Int,
+        @UserIdInt target: Int,
+    ) = intents.isNotEmpty()
+}
diff --git a/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java b/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java
index 0e4a6350..04ac1640 100644
--- a/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java
+++ b/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java
@@ -15,6 +15,8 @@
  */
 package com.android.intentresolver.emptystate;
 
+import static com.android.intentresolver.util.IntentUtils.isCrossProfileIntent;
+
 import android.annotation.NonNull;
 import android.annotation.UserIdInt;
 import android.app.AppGlobals;
@@ -22,8 +24,6 @@ import android.content.ContentResolver;
 import android.content.Intent;
 import android.content.pm.IPackageManager;
 
-import com.android.intentresolver.IntentForwarderActivity;
-
 import java.util.List;
 
 /**
@@ -52,7 +52,6 @@ public class CrossProfileIntentsChecker {
     public boolean hasCrossProfileIntents(
             List<Intent> intents, @UserIdInt int source, @UserIdInt int target) {
         return intents.stream().anyMatch(intent ->
-                null != IntentForwarderActivity.canForward(intent, source, target,
-                        mPackageManager, intent.resolveTypeIfNeeded(mContentResolver)));
+                isCrossProfileIntent(intent, source, target, mPackageManager, mContentResolver));
     }
 }
diff --git a/java/src/com/android/intentresolver/util/IntentUtils.kt b/java/src/com/android/intentresolver/util/IntentUtils.kt
index c20479c8..dccacf8d 100644
--- a/java/src/com/android/intentresolver/util/IntentUtils.kt
+++ b/java/src/com/android/intentresolver/util/IntentUtils.kt
@@ -18,20 +18,91 @@
 
 package com.android.intentresolver.util
 
+import android.app.AppGlobals
+import android.content.ContentResolver
 import android.content.Intent
+import android.content.pm.IPackageManager
+import android.os.UserHandle
+import com.android.intentresolver.IntentForwarderActivity
 
-fun sanitizePayloadIntents(intents: List<Intent>): List<Intent> =
-    intents.map { intent ->
-        Intent(intent).also { sanitized ->
-            sanitized.setPackage(null)
-            sanitized.setComponent(null)
-            sanitized.selector?.let {
-                sanitized.setSelector(
-                    Intent(it).apply {
-                        setPackage(null)
-                        setComponent(null)
+@JvmOverloads
+fun prepareCrossProfileIntents(
+    contentResolver: ContentResolver,
+    targetIntent: Intent,
+    intents: List<Intent>,
+    source: UserHandle,
+    target: UserHandle,
+    packageManagerProvider: () -> IPackageManager = { AppGlobals.getPackageManager() },
+    isCrossProfileIntent: (Intent, Int, Int, IPackageManager, ContentResolver) -> Boolean =
+        ::isCrossProfileIntent,
+): List<Intent> {
+    val packageManager = packageManagerProvider()
+    if (
+        !isCrossProfileIntent(
+            targetIntent,
+            source.identifier,
+            target.identifier,
+            packageManager,
+            contentResolver,
+        )
+    ) {
+        // If the target intent can't be forwarded, then we can't forward any intents The empty
+        // collections will be handled by the NoCrossProfileEmptyStateProvider.
+        return emptyList()
+    }
+
+    return sanitizePayloadIntents(intents) { intent ->
+        // The first item in the list is the target intent (but it is not necessarily the same
+        // object as the targetIntent passed in to this method). See the
+        // ChooserRequest.payloadIntents.
+        intents[0] == intent ||
+            isCrossProfileIntent.invoke(
+                intent,
+                source.identifier,
+                target.identifier,
+                packageManager,
+                contentResolver,
+            )
+    }
+}
+
+fun sanitizePayloadIntents(
+    intents: List<Intent>,
+    predicate: (Intent) -> Boolean = { true },
+): List<Intent> =
+    buildList(capacity = intents.size) {
+        for (intent in intents) {
+            if (predicate(intent)) {
+                add(
+                    Intent(intent).also { sanitized ->
+                        sanitized.setPackage(null)
+                        sanitized.setComponent(null)
+                        sanitized.selector?.let {
+                            sanitized.setSelector(
+                                Intent(it).apply {
+                                    setPackage(null)
+                                    setComponent(null)
+                                }
+                            )
+                        }
                     }
                 )
             }
         }
     }
+
+// TODO: move `IntentForwarderActivity#canForward()` and related methods here.
+fun isCrossProfileIntent(
+    intent: Intent,
+    sourceUserId: Int,
+    targetUserId: Int,
+    packageManager: IPackageManager,
+    contentResolver: ContentResolver,
+) =
+    IntentForwarderActivity.canForward(
+        intent,
+        sourceUserId,
+        targetUserId,
+        packageManager,
+        intent.resolveTypeIfNeeded(contentResolver),
+    ) != null
diff --git a/tests/unit/src/com/android/intentresolver/util/IntentUtilsTest.kt b/tests/unit/src/com/android/intentresolver/util/IntentUtilsTest.kt
index 8042b82e..b7757002 100644
--- a/tests/unit/src/com/android/intentresolver/util/IntentUtilsTest.kt
+++ b/tests/unit/src/com/android/intentresolver/util/IntentUtilsTest.kt
@@ -17,33 +17,41 @@
 package com.android.intentresolver.util
 
 import android.content.ComponentName
+import android.content.ContentResolver
 import android.content.Intent
 import android.content.Intent.ACTION_SEND
+import android.content.Intent.ACTION_VIEW
+import android.content.pm.IPackageManager
+import android.net.Uri
+import android.os.UserHandle
 import com.google.common.truth.Truth.assertThat
 import org.junit.Test
+import org.mockito.kotlin.mock
 
 class IntentUtilsTest {
+    private val mockContentResolver = mock<ContentResolver>()
+    private val mockPackageManager = mock<IPackageManager>()
+
+    private val sourceUser = UserHandle.of(0)
+    private val targetUser = UserHandle.of(10)
+
     @Test
-    fun test_sanitizePayloadIntents() {
+    fun test_sanitizePayloadIntents_defaultPredicate() {
         val intents =
             listOf(
                 Intent(ACTION_SEND).apply { setPackage("org.test.example") },
                 Intent(ACTION_SEND).apply {
-                    setComponent(
-                        ComponentName.unflattenFromString("org.test.example/.TestActivity")
-                    )
+                    component = ComponentName.unflattenFromString("org.test.example/.TestActivity")
                 },
                 Intent(ACTION_SEND).apply {
-                    setSelector(Intent(ACTION_SEND).apply { setPackage("org.test.example") })
+                    selector = Intent(ACTION_SEND).apply { setPackage("org.test.example") }
                 },
                 Intent(ACTION_SEND).apply {
-                    setSelector(
+                    selector =
                         Intent(ACTION_SEND).apply {
-                            setComponent(
+                            component =
                                 ComponentName.unflattenFromString("org.test.example/.TestActivity")
-                            )
                         }
-                    )
                 },
             )
 
@@ -52,11 +60,152 @@ class IntentUtilsTest {
         assertThat(sanitized).hasSize(intents.size)
         for (i in sanitized) {
             assertThat(i.getPackage()).isNull()
-            assertThat(i.getComponent()).isNull()
-            i.getSelector()?.let {
+            assertThat(i.component).isNull()
+            i.selector?.let {
                 assertThat(it.getPackage()).isNull()
-                assertThat(it.getComponent()).isNull()
+                assertThat(it.component).isNull()
+            }
+        }
+    }
+
+    @Test
+    fun test_sanitizePayloadIntents_emptyList() {
+        val sanitized = sanitizePayloadIntents(emptyList())
+        assertThat(sanitized).isEmpty()
+    }
+
+    @Test
+    fun test_sanitizePayloadIntents_withPredicate() {
+        val intentToKeep = Intent(ACTION_SEND).apply { type = "text/plain" }
+        val intentToFilter = Intent(ACTION_VIEW).apply { data = Uri.parse("http://example.com") }
+        val intents = listOf(intentToKeep, intentToFilter)
+
+        val sanitized = sanitizePayloadIntents(intents) { it.action == ACTION_SEND }
+
+        assertThat(sanitized).hasSize(1)
+        assertThat(sanitized[0].action).isEqualTo(ACTION_SEND)
+    }
+
+    @Test
+    fun prepareCrossProfileIntents_targetIntentCannotBeForwarded_returnsEmptyList() {
+        val targetIntent = Intent(ACTION_SEND).apply { type = "text/plain" }
+        val payloadIntents = listOf(targetIntent)
+
+        val result =
+            prepareCrossProfileIntents(
+                mockContentResolver,
+                targetIntent,
+                payloadIntents,
+                sourceUser,
+                targetUser,
+                { mockPackageManager },
+                { _, _, _, _, _ -> false },
+            )
+
+        assertThat(result).isEmpty()
+    }
+
+    @Test
+    fun prepareCrossProfileIntents_somePayloadsCannotBeForwarded_returnsFilteredList() {
+        val targetIntent = Intent(ACTION_SEND).apply { type = "text/plain" }
+        val forwardablePayload = Intent(ACTION_SEND).apply { type = "image/png" }
+        val nonForwardablePayload = Intent(ACTION_VIEW)
+        val intents = listOf(targetIntent, forwardablePayload, nonForwardablePayload)
+
+        val result =
+            prepareCrossProfileIntents(
+                mockContentResolver,
+                targetIntent,
+                intents,
+                sourceUser,
+                targetUser,
+                { mockPackageManager },
+                { intent, _, _, _, _ ->
+                    when (intent) {
+                        targetIntent -> true
+                        forwardablePayload -> true
+                        else -> false
+                    }
+                },
+            )
+
+        assertThat(result).hasSize(2)
+        assertThat(result[0].action).isEqualTo(ACTION_SEND)
+        assertThat(result[0].type).isEqualTo("text/plain")
+        assertThat(result[1].action).isEqualTo(ACTION_SEND)
+        assertThat(result[1].type).isEqualTo("image/png")
+    }
+
+    @Test
+    fun prepareCrossProfileIntents_allPayloadsCanBeForwarded_returnsSanitizedList() {
+        val targetIntent =
+            Intent(ACTION_SEND).apply {
+                type = "text/plain"
+                setPackage("foo")
             }
+        val payloadIntent =
+            Intent(ACTION_SEND).apply {
+                type = "image/png"
+                component = ComponentName("bar", "baz")
+            }
+        val intents = listOf(targetIntent, payloadIntent)
+
+        val result =
+            prepareCrossProfileIntents(
+                mockContentResolver,
+                targetIntent,
+                intents,
+                sourceUser,
+                targetUser,
+                { mockPackageManager },
+                { _, _, _, _, _ -> true },
+            )
+
+        assertThat(result).hasSize(2)
+        result.forEach {
+            assertThat(it.getPackage()).isNull()
+            assertThat(it.component).isNull()
+        }
+    }
+
+    @Test
+    fun prepareCrossProfileIntents_emptyIntentList_returnsEmptyList() {
+        val targetIntent = Intent(ACTION_SEND).apply { type = "text/plain" }
+
+        val result =
+            prepareCrossProfileIntents(
+                mockContentResolver,
+                targetIntent,
+                emptyList(),
+                sourceUser,
+                targetUser,
+                { mockPackageManager },
+                { _, _, _, _, _ -> true },
+            )
+
+        assertThat(result).isEmpty()
+    }
+
+    @Test
+    fun prepareCrossProfileIntents_usesPackageManagerProvider() {
+        val targetIntent = Intent(ACTION_SEND).apply { type = "text/plain" }
+        val intents = listOf(targetIntent)
+        var providerCalled = false
+        val packageManagerProvider = {
+            providerCalled = true
+            mockPackageManager
         }
+
+        prepareCrossProfileIntents(
+            mockContentResolver,
+            targetIntent,
+            intents,
+            sourceUser,
+            targetUser,
+            packageManagerProvider,
+            { _, _, _, _, _ -> true },
+        )
+
+        assertThat(providerCalled).isTrue()
     }
 }
-- 
2.55.0.897.gb25b4bd76c-goog

